src/puya/ir/optimize/compiled_reference.py

```diff
@@ -0,0 +1,163 @@
+"""
+
+from collections.abc import Mapping
+
+import attrs
+from immutabledict import immutabledict
+
+from puya import algo_constants, log
+from puya.algo_constants import HASH_PREFIX_PROGRAM, MAX_APP_PAGE_SIZE, MAX_BYTES_LENGTH
+from puya.awst.txn_fields import TxnField
+from puya.errors import InternalError
+from puya.ir import models as ir
+from puya.ir.optimize.context import IROptimizeContext
+from puya.ir.types_ import AVMBytesEncoding
+from puya.ir.visitor_mutator import IRMutator
+from puya.models import (
+    TemplateValue,
+)
+from puya.utils import Address, sha512_256_hash
+
+logger = log.get_logger(__name__)
+
+
+def replace_compiled_references(context: IROptimizeContext, subroutine: ir.Subroutine) -> bool:
+    replacer = CompiledReferenceReplacer(context)
+    for block in subroutine.body:
+        replacer.visit_block(block)
+    return replacer.modified
+
+
+@attrs.define
+class CompiledReferenceReplacer(IRMutator):
+    context: IROptimizeContext
+    modified: bool = False
+
+    def visit_compiled_logicsig_reference(  # type: ignore[override]
+        self,
+        const: ir.CompiledLogicSigReference,
+    ) -> ir.CompiledLogicSigReference | ir.Constant:
+        if not _is_constant(const.template_variables):
+            return const
+        template_constants = _get_template_constants(
+            self.context.options.template_variables, const.template_variables
+        )
+        program_bytecode = self.context.get_program_bytecode(
+            const.artifact, "logic_sig", template_constants
+        )
+        address_public_key = sha512_256_hash(HASH_PREFIX_PROGRAM + program_bytecode)
+        return ir.AddressConstant(
+            value=Address.from_public_key(address_public_key).address,
+            source_location=const.source_location,
+        )
+
+    def visit_compiled_contract_reference(  # type: ignore[override]
+        self,
+        const: ir.CompiledContractReference,
+    ) -> ir.CompiledContractReference | ir.Constant:
+        field = const.field
+        if field in (
+            TxnField.GlobalNumUint,
+            TxnField.GlobalNumByteSlice,
+            TxnField.LocalNumUint,
+            TxnField.LocalNumByteSlice,
+        ):
+            try:
+                state_total = self.context.state_totals[const.artifact]
+            except KeyError:
+                raise InternalError(
+                    f"Invalid contract reference: {const.artifact}", const.source_location
+                ) from None
+            match field:
+                case TxnField.GlobalNumUint:
+                    total = state_total.global_uints
+                case TxnField.GlobalNumByteSlice:
+                    total = state_total.global_bytes
+                case TxnField.LocalNumUint:
+                    total = state_total.local_uints
+                case TxnField.LocalNumByteSlice:
+                    total = state_total.local_bytes
+                case _:
+                    raise InternalError(
```

Comment:
annoying that mypy can't narrow the type here to allow `typing.assert_never(field)`